Document MenuBar and ContextMenu (#7502)
slint-ui/slint: docs/astro/src/content/docs/reference/window/contextmenu.mdx

---
<!-- Copyright © SixtyFPS GmbH <info@slint.dev> ; SPDX-License-Identifier: MIT -->

> title: Window
> description: Window element api.

Reviewer: I was wondering why this didn't show up in my local build...

```suggestion
title: ContextMenu
description: ContextMenu element api.
```

---
import SlintProperty from '/src/components/SlintProperty.astro';
import Link from '/src/components/Link.astro';

> `ContextMenu` is an invisible element that is used to show a context menu.

Reviewer: Using "invisible" and "show" in the same sentence like that is weird IMO :-).

```suggestion
Use the non-visual `ContextMenu` element to declare a context menu.
```

The context menu is shown if the user right-clicks on the area covered by the `ContextMenu` element,

> or if the user presses the "Menu" key on their keyboard while a `FocusScope` within the `ContextMenu` is focused.

Reviewer: (suggesting to use the same term as the name of the property)

```suggestion
or if the user presses the "Menu" key on their keyboard while a `FocusScope` within the `ContextMenu` has focus.
```

> It is also possible to show the context menu by calling the `show` function on the `ContextMenu` element.

Reviewer: ```suggestion
Call the `show()` function on the `ContextMenu` element to programmatically show the context menu.
```

> In addition to normal sub elements, the `ContextMenu` element can contain `MenuItem` elements.
> The MenuItem elements are defining the actual menu.

Reviewer: I think it's better to get straight to the point than to start about an unusual case (somebody placing an `Image` as a child).

```suggestion
Define the structure of the menu by placing `MenuItem` elements into the `ContextMenu`.
```

## Function

### show(Point)

> Call this function to programmatically show the context menu at the given position relative to the `ContextMenu` element.

Reviewer: I have the feeling that we might revisit these semantics in the API review ;-)

Author: ok, feel free to let me know what feels wrong there.

## `MenuItem`

A `MenuItem` represents a single menu entry. It can be placed as a child of a `ContextMenu`,
<Link type="MenuBar" />, or another `MenuItem`.

### Properties of `MenuItem`

#### title
<SlintProperty propName="title" typeName="string" defaultValue='""'>
The label of the entry within the menu.
</SlintProperty>

### Callbacks of `MenuItem`

#### activated()

Invoked when the menu entry is activated.

> This is only called if the menu entry doesn't have submenus

Reviewer: ```suggestion

:::note{Note}
This is only called if the menu entry doesn't have submenus.
:::
```

## Example

```slint
export component Example {

>     ContextMenu {
>         MenuItem {
>             title: "Copy";
>             activated => { debug("Copy"); }
>         }
>         MenuItem {
>             title: "Paste";
>             activated => { debug("Paste"); }
>         }
>         MenuItem {
>             title: "Cut";
>             activated => { debug("Cut"); }
>         }
>     }

Reviewer: Nitpick: Best practices to use `@tr()` in examples where we'd also recommend the use of it, and fix order to match what's used in real life :-)

```suggestion
    ContextMenu {
        MenuItem {
            title: @tr("Cut");
            activated => { debug("Cut"); }
        }
        MenuItem {
            title: @tr("Copy");
            activated => { debug("Copy"); }
        }
        MenuItem {
            title: @tr("Paste");
            activated => { debug("Paste"); }
        }
    }
```
